File: CP/code/chap.2/uva146.java
import java.util.*;
import java.io.*;

class Main {
    private static String readCode(BufferedReader sin) throws IOException {
        String temp = sin.readLine().trim();
        while (temp.isEmpty())
            temp = sin.readLine().trim();

        return temp;
   }

   private static <E extends Comparable<? super E>> boolean nextPermutation(List<E> perm) {
       int k = perm.size() - 2;
       while (k >= 0) {
           if (perm.get(k).compareTo(perm.get(k+1)) < 0)
               break;
           k--;
       }
       if (k < 0)
           return false;

       int l = perm.size() - 1;
       while (l > k) {
           if (perm.get(l).compareTo(perm.get(k)) > 0)
               break;
           l--;
       }

       E temp = perm.get(l);
       perm.set(l, perm.get(k));
       perm.set(k, temp);

       Collections.reverse(perm.subList(k+1, perm.size()));

       return true;
    }

   public static void main(String[] args) throws IOException {
       try (BufferedReader sin = new BufferedReader(new InputStreamReader(System.in));
            PrintWriter sout = new PrintWriter(System.out)) {
           while (true) {
               String code = readCode(sin);
               if (code == null || code.equals("#"))
                   break;

               ArrayList<String> codeList = new ArrayList<String>(Arrays.asList(code.split("")));
               if (nextPermutation(codeList)) {
                   String successor = String.join("", codeList);
                   sout.print(successor + System.lineSeparator());
               } else
                       sout.print("No Successor" + System.lineSeparator());
           }
       }
   } 
}

File: CP/code/chap.2/uva146.cs
using System;
using System.Linq;
using System.IO;
using System.Collections.Generic;

class UVA146
{
    private static string ReadCode(StreamReader fin)
    {
        var temp = fin.ReadLine().Trim();
        while (temp.Length == 0)
            temp = fin.ReadLine().Trim();

        return temp;
    }

    private static bool NextPermutation<T>(List<T> perm)
        where T : IComparable<T>
    {
        int k = perm.Count - 2;
        while (k >= 0)
        {
            if (perm[k].CompareTo(perm[k+1]) < 0)
                break;
            k--;
        }
        if (k < 0)
            return false;

        int l = perm.Count - 1;
        while (l > k)
        {
            if (perm[l].CompareTo(perm[k]) > 0)
                break;
            l--;
        }
        T temp = perm[l];
        perm[l] = perm[k];
        perm[k] = temp;

        perm.Reverse(k+1, perm.Count-k-1);

        return true;
    }



    public static void Main()
    {
        using (var cin = new StreamReader(Console.OpenStandardInput()))
            using (var cout = new StreamWriter(Console.OpenStandardOutput()))
            {
                while (true)
                {
                    string code = ReadCode(cin);
                    if (code == null || code == "#")
                        break;
                    var codeSeq = code.ToCharArray().ToList();
                    if (NextPermutation(codeSeq))
                    {
                        String successor = String.Join("", codeSeq);
                        cout.WriteLine(successor);
                    }
                    else
                    {
                        cout.WriteLine("No Successor");
                    }

                }

            }
    }
}
